Support Vaadin 14.4.3+

We found that `SelectionTreeGrid` throws an exception at runtime if we use Vaadin 14.4.3+. Looks like the `HierarchyMapper` has been updated in the flow-data library. Are there any plans to support the new Vaadin versions?

```
Caused by: java.lang.NoSuchFieldException: expandedItemIds
	at java.lang.Class.getDeclaredField(Class.java:2070)
	at com.vaadin.componentfactory.selectiongrid.SelectionTreeGrid.<init>(SelectionTreeGrid.java:71)
```

--- a/pom.xml
+++ b/pom.xml
@@ -3,7 +3,7 @@
 
     <groupId>com.vaadin.componentfactory</groupId>
     <artifactId>selection-grid-flow-root</artifactId>
-    <version>0.2.0</version>
+    <version>0.3.0</version>
     <packaging>pom</packaging>
     <modules>
         <module>selection-grid-flow</module>
@@ -13,12 +13,12 @@
     <description>selection-grid-flow</description>
     
     <properties>
-        <vaadin.version>14.3.8</vaadin.version>
+        <vaadin.version>14.4.4</vaadin.version>
         <maven.compiler.source>1.8</maven.compiler.source>
         <maven.compiler.target>1.8</maven.compiler.target>
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
         <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
-        <selection-grid-flow.version>0.2.0</selection-grid-flow.version>
+        <selection-grid-flow.version>0.3.0</selection-grid-flow.version>
     </properties>
     <inceptionYear>2020</inceptionYear>
     <organization>

--- a/selection-grid-flow-demo/pom.xml
+++ b/selection-grid-flow-demo/pom.xml
@@ -6,7 +6,7 @@
 
     <groupId>com.vaadin.componentfactory</groupId>
     <artifactId>selection-grid-flow-demo</artifactId>
-    <version>0.2.0</version>
+    <version>0.3.0</version>
 
     <name>Selection Grid Demo</name>
     <packaging>war</packaging>
@@ -18,7 +18,7 @@
     </organization>
 
     <properties>
-        <vaadin.version>14.3.8</vaadin.version>
+        <vaadin.version>14.4.4</vaadin.version>
         <maven.compiler.source>1.8</maven.compiler.source>
         <maven.compiler.target>1.8</maven.compiler.target>
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
@@ -106,7 +106,7 @@
        <dependency>
            <groupId>com.vaadin.componentfactory</groupId>
            <artifactId>selection-grid-flow</artifactId>
-           <version>0.2.0</version>
+           <version>0.3.0</version>
        </dependency>
 
         <dependency>

--- a/selection-grid-flow/pom.xml
+++ b/selection-grid-flow/pom.xml
@@ -6,7 +6,7 @@
 
     <groupId>com.vaadin.componentfactory</groupId>
     <artifactId>selection-grid-flow</artifactId>
-    <version>0.2.0</version>
+    <version>0.3.0</version>
     <packaging>jar</packaging>
 
     <name>Selection Grid</name>
@@ -19,7 +19,7 @@
     </organization>
 
     <properties>
-        <vaadin.version>14.3.8</vaadin.version>
+        <vaadin.version>14.4.4</vaadin.version>
         <maven.compiler.source>1.8</maven.compiler.source>
         <maven.compiler.target>1.8</maven.compiler.target>
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>

--- a/selection-grid-flow/src/main/java/com/vaadin/componentfactory/selectiongrid/SelectionTreeGrid.java
+++ b/selection-grid-flow/src/main/java/com/vaadin/componentfactory/selectiongrid/SelectionTreeGrid.java
@@ -28,23 +28,16 @@ import com.vaadin.flow.component.grid.GridMultiSelectionModel;
 import com.vaadin.flow.component.grid.GridSelectionModel;
 import com.vaadin.flow.component.treegrid.TreeGrid;
 import com.vaadin.flow.data.provider.DataProvider;
-import com.vaadin.flow.data.provider.KeyMapper;
-import com.vaadin.flow.data.provider.hierarchy.HierarchicalArrayUpdater;
 import com.vaadin.flow.data.provider.hierarchy.HierarchicalDataCommunicator;
 import com.vaadin.flow.data.provider.hierarchy.HierarchyMapper;
 import com.vaadin.flow.data.provider.hierarchy.TreeDataProvider;
 import com.vaadin.flow.data.renderer.TemplateRenderer;
 import com.vaadin.flow.function.SerializableComparator;
 import com.vaadin.flow.function.ValueProvider;
-import com.vaadin.flow.internal.JsonUtils;
 import com.vaadin.flow.internal.Range;
-import elemental.json.Json;
-import elemental.json.JsonObject;
 
-import java.lang.reflect.Field;
 import java.lang.reflect.Method;
 import java.util.ArrayList;
-import java.util.HashMap;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Set;
@@ -55,58 +48,6 @@ import java.util.stream.Collectors;
 @JsModule("./src/vcf-selection-grid.js")
 @JsModule("./src/selection-grid.js")
 public class SelectionTreeGrid<T> extends TreeGrid<T> {
-    /// TEMPORARY FIX FOR https://github.com/vaadin/vaadin-grid/issues/1820
-    /// Remove this once this issue is closed and released
-    {
-        // Get fields through reflection
-        final Field hierarchyMapperField;
-        final Field expandedItemIdsField;
-        final Field objectIdKeyMapField;
-        try {
-            objectIdKeyMapField = KeyMapper.class
-                .getDeclaredField("objectIdKeyMap");
-            hierarchyMapperField = HierarchicalDataCommunicator.class
-                .getDeclaredField("mapper");
-            expandedItemIdsField = HierarchyMapper.class
-                .getDeclaredField("expandedItemIds");
-
-            expandedItemIdsField.setAccessible(true);
-            hierarchyMapperField.setAccessible(true);
-            objectIdKeyMapField.setAccessible(true);
-        } catch (NoSuchFieldException e) {
-            throw new RuntimeException(e);
-        }
-
-        // Create attach listener
-        addAttachListener(event -> {
-            try {
-                HierarchicalDataCommunicator<T> communicator = getDataCommunicator();
-                HierarchyMapper mapper = (HierarchyMapper) hierarchyMapperField
-                    .get(communicator);
-                Set<Object> expandedItemIds = (Set<Object>) expandedItemIdsField
-                    .get(mapper);
-                HashMap<Object, String> objectIdKeyMap = (HashMap) objectIdKeyMapField
-                    .get(communicator.getKeyMapper());
-
-                HierarchicalArrayUpdater.HierarchicalUpdate update = (HierarchicalArrayUpdater.HierarchicalUpdate) getArrayUpdater()
-                    .startUpdate(mapper.getRootSize());
-                update.enqueue("$connector.expandItems",
-                    expandedItemIds
-                        .stream()
-                        .map(objectIdKeyMap::get)
-                        .map(key -> {
-                            JsonObject json = Json.createObject();
-                            json.put("key", key);
-                            return json;
-                        }).collect(
-                        JsonUtils.asArray()));
-                event.getUI().beforeClientResponse(this, ctx -> update.commit());
-            } catch (IllegalAccessException e) {
-                throw new RuntimeException(e);
-            }
-        });
-    }
-    /// END TEMPORARY FIX
     /**
      * Focus on the first cell on the row
      *

--- a/selection-grid-flow/src/main/resources/META-INF/resources/frontend/src/vcf-selection-grid.js
+++ b/selection-grid-flow/src/main/resources/META-INF/resources/frontend/src/vcf-selection-grid.js
@@ -60,26 +60,26 @@ class VcfSelectionGridElement extends ElementMixin(ThemableMixin(GridElement)) {
 
     }
 
-
-    focusOnCell(rowNumber, cellNumber) {
-        if (rowNumber < 0 || cellNumber < 0) {
-            throw "index out of bound";
+    focusOnCell(rowNumber, cellNumber, nbOfCalls = 1) {
+        if (nbOfCalls < 11) { // dont make an infinite loop
+            if (rowNumber < 0 || cellNumber < 0) {
+                throw "index out of bound";
+            }
+            this.scrollToIndex(rowNumber);
+            /** workaround when the expanded node opens children the index is outside the grid size
+             * https://github.com/vaadin/vaadin-grid/issues/2060
+             * Remove this once this is fixed
+             **/
+            if (rowNumber > this._effectiveSize) {
+                const that = this;
+                setTimeout(() => {
+                    that.focusOnCell(rowNumber, cellNumber, nbOfCalls + 1);
+                }, 200);
+            } else {
+                this._startToFocus(rowNumber, cellNumber);
+            }
+            /** End of workaround **/
         }
-        this.scrollToIndex(rowNumber);
-        /** workaround when the expanded node opens children the index is outside the grid size
-         * https://github.com/vaadin/vaadin-grid/issues/2060
-         * Remove this once this is fixed
-         **/
-        if (rowNumber > this._effectiveSize) {
-            const that = this;
-            setTimeout(() => {
-                that.scrollToIndex(rowNumber);
-                that._startToFocus(rowNumber, cellNumber);
-            }, 200);
-        } else {
-            this._startToFocus(rowNumber, cellNumber);
-        }
-        /** End of workaround **/
     };
 
     _startToFocus(rowNumber, cellNumber) {

